### ISSUE TITLE
Series index is not retained after sorting values

### ISSUE THREAD

**Issue by Emporea** [javascriptdata/danfojs#357]
sortValues seems inconvinient because it does reset the index and its hard to reindex it if I only have numbers...

The Index gets ignored too often in general for example:

Series.toJSON() or toCSV() only includes the values. This is inconvinient. There should be an option to not drop the index.
Pandas doesn't drop the index either.

Also is it possible to combine the functions valueCounts() and sortValues, just like pandas does? The missing index when passing a Series makes further calculations unnecessarily complicated..

If it is intended to leave a passed Series 1 dimensional there should be atleast a method that converts a Series to a Dataframe by adding a new Index. This is done by reindex() in pandas, but in danfojs this only seems to drop the current index and give it new values.

**Comment by Emporea** [comment]
A workaround for me would be to replace line 148 in src/danfojs-base/io/node/io.json.ts with
`obj[df.columns[0]] = Object.fromEntries(df.index.map((e, i) => [e, df.values[i]]))`

**Comment by risenW** [comment]
@Emporea Thanks for the PR. I will address each paragraph:

> sortValues seems inconvenient because it does reset the index and its hard to reindex it if I only have numbers...

The intended behavior of `sortValues` is to sort each value position in the Series, and as such, the index is updated with it. This is similar to the behavior in Pandas. See https://pandas.pydata.org/docs/reference/api/pandas.Series.sort_values.html

> Series.toJSON() or toCSV() only includes the values. This is inconvinient. There should be an option to not drop the index.
Pandas doesn't drop the index either.

This is a good call and can be added. If you want to send a PR for this, I'll happy to review.

> Also is it possible to combine the functions valueCounts() and sortValues, just like pandas does? The missing index when passing a Series makes further calculations unnecessarily complicated..

In general, most method on Series and DataFrame are chainable since they return a Series/DataFrame. So this works:
```javascript
import { Series } from "danfojs-node"

let sf = new Series([1, 1, 2, 3, 6, 7, 8, 7, 7, 9, 0, 0, 3])
sf.valueCounts().sortValues().print()
//output
```
<img width="248" alt="Screen Shot 2022-01-18 at 7 49 14 AM" src="https://user-images.githubusercontent.com/29900845/149885237-4e5da3f8-465e-439a-a541-0790c9f96457.png">

>If it is intended to leave a passed Series 1 dimensional there should be atleast a method that converts a Series to a Dataframe by adding a new Index. This is done by reindex() in pandas, but in danfojs this only seems to drop the current index and give it new values.

This is a new feature request and welcome addition. We currently don't have a `reindex` function, just a `resetIndex` which does the exact thing it's meant for.

In conclusion, thanks for taking out time to suggest these. If you would like to work on any of the new features you have noted, do let me know, and I can assign you to this.

**Comment by Emporea** [comment]
Thanks for your answer!
I am very very new to javascript, I can see what I can do! But no pressure, I am still learning.



> The intended behavior of sortValues is to sort each value position in the Series, and as such, the index is updated with it. This is similar to the behavior in Pandas. See https://pandas.pydata.org/docs/reference/api/pandas.Series.sort_values.html

The difference is that pandas keeps the index, which is very useful.

Example:

```
var fruits = ["apple", "apple", "apple", "banana", "grape", "orange", "orange"];
var testdf = new danfo.Series(fruits);
testdf.valueCounts()

$data': [ 3, 1, 1, 2 ]
```

while pandas does this:
```
fruits = ["apple", "apple", "apple", "banana", "grape", "orange", "orange"]
testdf = pd.Series(fruits)
testdf.value_counts()

apple     3
orange    2
grape     1
banana    1
```

I know I can chain sort_values. But in the end with danfo I dont even know which item occures how often because of the dropped index in all Series operations.

> This is a new feature request and welcome addition. We currently don't have a reindex function, just a resetIndex which does the exact thing it's meant for.
The pandas reset_index() has an argument "drop" which is false by default. This means that a pd.Series.reset_index() actually creates a Dataframe and takes the old index as a new column.

![image](https://user-images.githubusercontent.com/21969523/149903478-9a3c5396-3854-4f8b-97f4-d4f45a753d69.png)


Needless to say danfo-js is a very nice project and I appreciate having this pandas-feel in this new environment for me! And even tho I dont have the self-confidence and knowledge to contribute to this project I will see what I can do. :)

**Comment by risenW** [comment]
> Thanks for your answer! I am very very new to javascript, I can see what I can do! But no pressure, I am still learning.
> 
> > The intended behavior of sortValues is to sort each value position in the Series, and as such, the index is updated with it. This is similar to the behavior in Pandas. See https://pandas.pydata.org/docs/reference/api/pandas.Series.sort_values.html
> 
> The difference is that pandas keeps the index, which is very useful.
> 
> Example:
> 
> ```
> var fruits = ["apple", "apple", "apple", "banana", "grape", "orange", "orange"];
> var testdf = new danfo.Series(fruits);
> testdf.valueCounts()
> 
> $data': [ 3, 1, 1, 2 ]
> ```
> 
> while pandas does this:
> 
> ```
> fruits = ["apple", "apple", "apple", "banana", "grape", "orange", "orange"]
> testdf = pd.Series(fruits)
> testdf.value_counts()
> 
> apple     3
> orange    2
> grape     1
> banana    1
> ```
> 
> I know I can chain sort_values. But in the end with danfo I dont even know which item occures how often because of the dropped index in all Series operations.
> 
> > This is a new feature request and welcome addition. We currently don't have a reindex function, just a resetIndex which does the exact thing it's meant for.
> 
> The pandas reset_index() has an argument "drop" which is false by default. This means that a pd.Series.reset_index() actually creates a Dataframe and takes the old index as a new column.
> 
> ![image](https://user-images.githubusercontent.com/21969523/149903478-9a3c5396-3854-4f8b-97f4-d4f45a753d69.png)
> 
> Needless to say danfo-js is a very nice project and I appreciate having this pandas-feel in this new environment for me! And even tho I dont have the self-confidence and knowledge to contribute to this project I will see what I can do. :)


No pressure, I understand your request now, at least regarding the `resetIndex`. Also, I don't know if you this, but you can print out the content of a DataFrame or Series by calling `.print()`:

```javascript
var fruits = ["apple", "apple", "apple", "banana", "grape", "orange", "orange"];
var testdf = new danfo.Series(fruits);
testdf.valueCounts().print() //gives you a well formated table in console
testdf.valueCounts().values //returns the data in array format


```

**Comment by Emporea** [comment]
Yes ofc I know about print().

Whats sortValues missing is the preservation of the index, it has to be also sorted regarding its sorted values.
We need an equivalent of [this line in pandas](https://github.com/pandas-dev/pandas/blob/a474af5b4526dc333f62260587a1d8ef494df57c/pandas/core/series.py#L3524):
Or to be precice this extended numpy.argsort function in pandas called [nargsort
](https://github.com/pandas-dev/pandas/blob/a474af5b4526dc333f62260587a1d8ef494df57c/pandas/core/sorting.py#L351)

**Comment by risenW** [comment]
> ```js
> let sf = new Series([1, 1, 2, 3, 6, 7, 8, 7, 7, 9, 0, 0, 3])
> sf.valueCounts().sortValues().print()
> ```

Hey, @Emporea I still don't understand what you're trying to achieve. Can you give me an example of the result you want and the current result Danfojs is giving you?

**Comment by Emporea** [comment]
My node danfo.js@1.0.1 npm files seem not to be the same as these git files.


I modified [toJSON()](https://github.com/javascriptdata/danfojs/blob/52ee69da50fd36161057f701f426e7b618e37ab1/src/danfojs-base/io/node/io.json.ts#L143) so that it doesnt output it inside a empty '0' column.

```
var $toJSON = function (df, options) {
    var _a = __assign({ filePath: undefined, format: "column" }, options), filePath = _a.filePath, format = _a.format;
    if (df.$isSeries) {
        var obj = {};
        obj = Object.fromEntries(df.index.map((e, i) => [e, df.values[i]]))        // I changed obj[df.columns[0]] to obj
        if (filePath) {
            if (!filePath.endsWith(".json")) {
                filePath = filePath + ".json";
            }
            fs_1.default.writeFileSync(filePath, JSON.stringify(obj));
        }
        else {
            return obj;
        }
    }
```

```
before_change= {
  0: {
    wikipedia: 2,
    None: 1,
    github: 1,
  },
};

after_change= {
  wikipedia: 2,
  None: 1,
  github: 1,
};
```

> A workaround for me would be to replace line 148 in src/danfojs-base/io/node/io.json.ts with `obj[df.columns[0]] = Object.fromEntries(df.index.map((e, i) => [e, df.values[i]]))`

this is obselete now.

**Comment by Emporea** [comment]
> Hey, @Emporea I still don't understand what you're trying to achieve. Can you give me an example of the result you want and the current result Danfojs is giving you?

```
  let index = ["apple", "banana", "orange", "grape"];
  let data = [3, 6, 2, 9];

  let sf = new danfo.Series(data, { index });
  sf.sortValues().print();
```
without change gives me:
![image](https://user-images.githubusercontent.com/21969523/150126703-065c9240-72d5-4a5b-9782-482e26592e9e.png)

but with the change of adding [real_SortedIdx](https://github.com/javascriptdata/danfojs/issues/357#issuecomment-1016270290) I get
![image](https://user-images.githubusercontent.com/21969523/150126840-107a5cd6-9043-4e4f-8896-da0c6c42471e.png)

The index got preserved and is not just a integer. Do you understand?

**Comment by risenW** [comment]
> let index = ["apple", "banana", "orange", "grape"];
>   let value = [3, 6, 2, 9];
> 
>   let sf = new danfo.Series(value, { index });
>   sf.sortValues().print();

Ah I get you now. This index is not retained for Series. It was retained in DataFrame. I'll get this fix now.

**Comment by risenW** [comment]
@Emporea I just pushed a new version (v1.0.2) which to fix this. The new version also adds the `at` and `iat` methods. Do test and let me know.

**Comment by Emporea** [comment]
Yeah, your fix works too! Now its official and I dont have to fix it manually :D

Could you take a look at the [toJSON()](https://github.com/javascriptdata/danfojs/issues/357#issuecomment-1016270290) function where I did a workaround to get rid of the unnecessary "0" dict? Should I create a new issue? Or is my fix (changing obj[df.columns[0]] to obj) sufficient for a PR?

**Comment by risenW** [comment]
> Yeah, your fix works too! Now its official and I dont have to fix it manually :D
> 
> Could you take a look at the [toJSON()](https://github.com/javascriptdata/danfojs/issues/357#issuecomment-1016270290) function where I did a workaround to get rid of the unnecessary "0" dict? Should I create a new issue? Or is my fix (changing obj[df.columns[0]] to obj) sufficient for a PR?

I think it should be sufficient, just add/update a couple of test in the toJSON and send a PR for review.

**Comment by Emporea** [comment]
Okey I can make a pull request. I actually did changed 2 things.

Do you know a better way for this?
`Object.fromEntries(df.index.map((e, i) => [e, df.values[i]]))`

I changed it to this, because toJSON() only gave the values, not the keys. So this concats them. Do you understand?
